Exportation of the Participations of a Cycle

yanicolivier opened the pull request:
| Q                                                      | R
| ------------------------------------------ | -------------------------------------------
| Type of contribution ?                      | Features
| Tickets (_issues_) concerned               | Fixed #154

---

##### What have you changed ?
Added the possibility to export a report in csv format from the admin panel.

##### How did you change it ?
I have added an admin action to compile the hours of the volunteer


##### Verification :

**PR standard**
 - [X] Branches naming convention: 
     - `prefix-snake_case_description`
     - ex: `enhancement-create_new_cell`
     - ex: `fix-pep8_standard_on_cell_model`
 - [X] Fill in this automatically generated PR template
 - [X] Do not include issue numbers in the PR title
 - [x] Include screenshots and animated GIFs in your pull request whenever possible

**Code standard**
 - [X] This Pull-Request fully meets the requirements defined in the issue
 - [x] Add or modify the attached tests
 - [X] Follow the Python Styleguide
 - [X] Document new code based on the Documentation Styleguide
 - [X] Use I18N functions when you add some text

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,134 @@
+
+        print(response.status_code)
+
+        self.assertEquals(response.status_code, 200)
```
We definitively want to test the content of the file and not just the status code!

yanicolivier replied:
Yes added the actual text

RignonNoel commented on source/apiVolontaria/volunteer/admin.py:
```diff
@@ -88,7 +92,88 @@ class EventAdmin(admin.ModelAdmin):
 
 
-admin.site.register(models.Cycle)
+def cycle_admin_generate_participation_report(modeladmin, request, queryset):
```
Is it possible to put this function inside the admin class ? 
I would just like the file to remain readable with time and new features.

RignonNoel commented on source/apiVolontaria/volunteer/admin.py:
```diff
@@ -88,7 +92,88 @@ class EventAdmin(admin.ModelAdmin):
+            value['total_time']
+        ])
+
+    return response
```
Can we create a generic functions to create the data export and just use this function to call the generic function ?
So when we will want to use it through the API it will be easy.

yanicolivier replied:
I have made a function in Cycle to compile the data in a dict and then the admin receive it an manage it to make a csv

RignonNoel commented on source/apiVolontaria/volunteer/admin.py:
```diff
@@ -88,6 +92,53 @@ class EventAdmin(admin.ModelAdmin):
+
+    actions = ['generate_participation_report_csv', ]
+
+    def generate_participation_report_csv(self, request, queryset):
```
Can you add a documentation on this function since it's not really standard for the moment ?
```
def generate_participation_report_csv(self, request, queryset):
    """
    Little description
    :param request: blahblahblah 
    :param queryset: blahblahblah
    :return: the format/structure of the reponse you return
    """
```

yanicolivier replied:
Good Point

RignonNoel commented on source/apiVolontaria/volunteer/admin.py:
```diff
@@ -88,6 +92,61 @@ class EventAdmin(admin.ModelAdmin):
+        :param request: The request object of the admin
+        :param queryset: This is the admin panel queryset.
+        :return: This will return a CSV file
+        """
```
You have some typos in the documentation :
 - `from this a cycle`
 - `cycle ang compile`

yanicolivier replied:
Oups, corrected.

RignonNoel commented on source/apiVolontaria/volunteer/models.py:
```diff
@@ -44,6 +44,42 @@ def is_active(self):
     def __str__(self):
         return '{}'.format(self.name)
 
+    def generate_participation_report_data(self):
```
I think you will need the same documentation here to explain what you do and return.

yanicolivier replied:
Yes make sense, added.

RignonNoel commented on source/apiVolontaria/volunteer/admin.py:
```diff
@@ -88,6 +92,61 @@ class EventAdmin(admin.ModelAdmin):
+                value['first_name'],
+                value['last_name'],
+                value['email'],
+                value['total_time']
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/models.py:
```diff
@@ -48,6 +48,58 @@ def is_active(self):
+
+        # Select all the participation of this Cycle
+        participations = Participation.objects.filter(
+            event__in=events_in_cycle
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/models.py:
```diff
@@ -48,6 +48,58 @@ def is_active(self):
+            event__in=events_in_cycle
+        )
+
+        data_dict = {}
```
init of dict with type constructor

yanicolivier replied:
What is the good practice for this ?

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+        profile = Profile(
+            user=user,
+            phone='1234567890',
+            mobile='0987654321'
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+            event=self.event,
+            user=self.user,
+            standby=False,
+            presence_status='P'
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+            event=self.event2,
+            user=self.user,
+            standby=False,
+            presence_status='P'
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+        self.admin = CycleAdmin(Cycle, self.site)
+
+        self.profile = Profile.objects.filter(
+            user__pk=self.participation.user.pk
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+        admin_pages = [
+            "/admin/volunteer/",
+            "/admin/volunteer/cycle/",
+            ]
```
indent:
```
        admin_pages = [
            "/admin/volunteer/",
            "/admin/volunteer/cycle/",
        ]
```

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+
+        self.assertEquals(
+            response.content,
+            b'first_name,last_name,email,total_time\r\n,,,96\r\n'
```
indent at end of line

yanicolivier replied:
Not sure about this one

francisjeanneau replied:
I think he meant a comma at end of line.
ie:
```
my_dict = {
    item1: "value1",
    item2: "value2",
}

RignonNoel replied:
Ahah, yes i wanted to say "coma" as said @francisjeanneau :rofl:

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_admin_Cycle.py:
```diff
@@ -0,0 +1,195 @@
+            event=self.event3,
+            user=self.user,
+            standby=False,
+            presence_status='I'
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_model_Cycle.py:
```diff
@@ -78,3 +80,98 @@ def test_str_method(self):
+        test_data_error = {
+            'error':
+                'All of the Participations presence '
+                'status must be initialised.'
```
coma at end of line

RignonNoel commented on source/apiVolontaria/volunteer/tests/tests_model_Cycle.py:
```diff
@@ -78,3 +80,98 @@ def test_str_method(self):
+                'first_name': self.admin.first_name,
+                'last_name': self.admin.last_name,
+                'email': self.admin.email,
+                'total_time': 300
```
coma at end of line

yanicolivier commented:
Yes, I have added a validation to export only on cycle at the time.  I
would have to zip the files to have multiple export.  I have added messages
to the admin to explain this.  Also, if any of the Participation entries of
the current cycle is not at the 'P'(Present) or 'A'(Absent) status, the
report cannot be exported since the Participations entries is not completed.

Le lun. 26 mars 2018 à 10:45, Rignon Noël <notifications@github.com> a
écrit :

> *@RignonNoel* requested changes on this pull request.
>
> Code is good for me however i've a problem with the behavior :
>
> [image: screenshot-2018-3-26 sqlite viewer]
> <https://user-images.githubusercontent.com/12053720/37913430-d9eb92dc-30e2-11e8-8a2c-2cf356338c8a.png>
> [image: Uploading Screenshot-2018-3-26 Select cycle to change Django site
> admin.png…]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Volontaria/API-Volontaria/pull/162#pullrequestreview-106934009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVGbu5pNIlCoSlclopM5211cfkRBFv2Rks5tiP8igaJpZM4Swmdv>
> .
>

RignonNoel commented:
> Also, if any of the Participation entries of the current cycle is not at the 'P'(Present) or 'A'(Absent) status, the report cannot be exported since the Participations entries is not completed. 
> 
>There is also a admin message for this case, you can see it in your second screenshot in red. The other message appear at the same place.
> @yanicolivier 

OH Ok! I didn't understand it this way without your explanation: the actual message on my screenshot is `All of the Participations presence status must be initialized` and i think it's really misleading since we've a status `Initialisation` :rofl: 

```
    PRESENCE_STATUS = (
        ('I', _('Initialisation')),
        ('A', _('Absent')),
        ('P', _('Present')),
    )
```

It's still work so i approve this PR since it's not a feature we want in the long run.

Thanks for your contribution :+1:
